core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -120,34 +108,59 @@ private[kafka] abstract class SubSourceLogic[K, V, Msg](
     consumer ! KafkaConsumerActor.Internal.Stop
   }
 
-  class SubSourceStage(tp: TopicPartition, consumer: ActorRef) extends GraphStage[SourceShape[Msg]] { stage =>
+  @tailrec
+  private def pump(): Unit = {
+    if (buffer.nonEmpty && isAvailable(shape.out)) {
+      val (tp, remains) = buffer.dequeue
+      buffer = remains
+      push(shape.out, (tp, createSource(tp)))
+      pump()
+    }
+  }
+
+  class SubSourceStage(tp: TopicPartition, consumer: ActorRef) extends GraphStage[SourceShape[Msg]] {
+    stage =>
     val out = Outlet[Msg]("out")
     val shape = new SourceShape(out)
 
     override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
       new GraphStageLogic(shape) with PromiseControl {
         val shape = stage.shape
         val requestMessages = KafkaConsumerActor.Internal.RequestMessages(0, Set(tp))
-        var requested = false
+        val pauseMessages = KafkaConsumerActor.Internal.PauseMessages(0, Set(tp))
+        val unpauseMessages = KafkaConsumerActor.Internal.UnpauseMessages(0, Set(tp))
+
+        var paused = false
         var self: StageActor = _
-        var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+        var buffer: mutable.Queue[ConsumerRecord[K, V]] = mutable.Queue()
+
+        val bufferSize = settings.sourceBufferSize
+
+        def checkBufferAndPause(): Unit =
+          if (!paused && buffer.size >= bufferSize) {
+            consumer.tell(pauseMessages, self.ref)
+            paused = true
+          }
+
+        def checkBufferAndUnpause(): Unit =
+          if (paused && buffer.size < bufferSize) {
+            consumer.tell(unpauseMessages, self.ref)
+            paused = false
+          }
 
         override def preStart(): Unit = {
           subsourceStartedCB.invoke((tp, this))
           self = getStageActor {
             case (_, msg: KafkaConsumerActor.Internal.Messages[K, V]) =>
-              requested = false
-              // do not use simple ++ because of https://issues.scala-lang.org/browse/SI-9766
-              if (buffer.hasNext) {
-                buffer = buffer ++ msg.messages
-              }
-              else {
-                buffer = msg.messages
-              }
+              //              requested = false
+
+              buffer.enqueue(msg.messages.toSeq: _*)
+              checkBufferAndPause()
               pump()
             case (_, Terminated(ref)) if ref == consumer =>
               failStage(new Exception("Consumer actor terminated"))
           }
+          consumer.tell(requestMessages, self.ref)
```

Comment:
Could you explain a bit more what's the flow now? Previously `pump()` was repeatedly sending `requestMessages` to the consumer actor. Now this message is sent  in `preStart` and I can't see where else.